src/core/thread/mesh_forwarder.cpp

```diff
@@ -1696,17 +1697,22 @@ void MeshForwarder::HandleSentFrame(Mac::Frame &aFrame, ThreadError aError)
                 child->mIndirectSendInfo.mMessage = NULL;
             }
 
-            mSendMessage->ClearChildMask(mNetif.GetMle().GetChildIndex(*child));
+            childIndex = mNetif.GetMle().GetChildIndex(*child);
 
-            if ((child->mMode & Mle::ModeTlv::kModeRxOnWhenIdle) == 0)
+            if (mSendMessage->GetChildMask(childIndex))
             {
-                child->mQueuedIndirectMessageCnt--;
-                otLogDebgMac(GetInstance(), "Sent to child (0x%x), still queued message (%d)",
-                             child->mValid.mRloc16, child->mQueuedIndirectMessageCnt);
+                mSendMessage->ClearChildMask(childIndex);
 
-                if (child->mQueuedIndirectMessageCnt == 0)
+                if ((child->mMode & Mle::ModeTlv::kModeRxOnWhenIdle) == 0)
```

Comment:
I think we should remove this check to ensure the marked messages and `mQueuedIndirectMessageCnt` stay in sync.